shared/lib/estimators/kNeighborsRegressor.ts

```diff
@@ -0,0 +1,175 @@
+  //  {
+  //    const len = distances.length
+  //
+  //    const min = distances.reduce( (x, y) => Math.min(x, y) )
+  //
+  //    let sum = 0
+  //    for (let i = len; i-- > 0;) {
+  //      const di = Math.exp(min - distances[i])
+  //      sum += di
+  //      distances[i] = di
+  //    }
+  //
+  //    for (let i = len; i-- > 0;)
+  //      distances[i] /= sum
+  //
+  //    return distances
+  //  }
+}
+
+/**
+ * Wraps a weightFn with input and ouput checks.
+ *
+ * @param weightsFn The wrapped weight function.
+ * @returns `weightsFn` wrapped with checks/assertions.
+ */
+const checkedWeightsFn = (weightsFn: WeightsFn) => (distances: number[]) => {
+  if (distances.some((d) => isNaN(d) || d < 0))
+    throw new Error(
+      'KNeighborsRegressor: metric function returned invalid distance.'
+    )
+
+  const weights = weightsFn(distances)
+
+  if (weights.length !== distances.length)
+    throw new Error(
+      'KNeighborsRegressor: array returned by weights must have the same length than the input.'
+    )
+
+  let sum = 0
+  for (let i = weights.length; i-- > 0; ) {
+    const wi = weights[i]
+    if (!(0 <= wi && wi <= 1))
+      // <- handles NaN
+      throw new Error(
+        'KNeighborsRegressor: entries returned by weights must be in range [0,1].'
+      )
+    sum += wi
+  }
+  if (!(Math.abs(sum - 1) <= 1e-8))
+    // <- handles NaN
+    throw new Error(
+      'KNeighborsRegressor: entries returned by weights must sum up to 1.'
+    )
+
+  return weights
+}
+
+export interface KNeighborsRegressorParams extends KNeighborsBaseParams {
+  nNeighbors?: int
+  weights?: WeightsFn | keyof typeof WEIGHTS_FUNCTIONS
+}
+
+export class KNeighborsRegressor extends KNeighborsBase {
+  private _weightsFn: WeightsFn
+  private _nNeighbors: int
+
+  constructor(params: KNeighborsRegressorParams = {}) {
+    super(params)
+    const { nNeighbors = 5, weights = 'uniform' } = params
+
+    this._nNeighbors = nNeighbors | 0
+    if (this._nNeighbors <= 0 || this._nNeighbors != nNeighbors)
+      throw new Error(
+        `new KNeighborsRegressor({nNeighbors}): nNeighbors must be a positive integer.`
+      )
+
+    if ('function' === typeof weights)
+      this._weightsFn = checkedWeightsFn(weights)
+    // <- TODO: remove check after testing
+    else if (Object.prototype.hasOwnProperty.call(WEIGHTS_FUNCTIONS, weights))
+      this._weightsFn = checkedWeightsFn(
+        WEIGHTS_FUNCTIONS[weights as keyof typeof WEIGHTS_FUNCTIONS]
+      )
+    else
+      throw new Error(
+        'new KNeighborsRegressor({weights}): weights must be of type ' +
+          Object.keys(WEIGHTS_FUNCTIONS).join(' | ') +
+          ' | (distance: number[n <= nNeighbors]) => number[n].'
+      )
+  }
+
+  async fit(X: Scikit2D, y: Scikit1D) {
+    await super._fit(X, y)
+    return this
+  }
+
+  predict<T extends Scikit1D | Scikit2D>(
+    _X: T
+  ): T extends Scikit1D
+    ? Scalar
+    : T extends Scikit2D
+    ? Tensor1D
+    : Scalar | Tensor1D {
```

Comment:
I'm curious why we would need this predict function to be this general. What's stopping us from just going

```
predict(_X: Scikit1D | Scikit2D): Tensor1D {
   // code here
}
```